relations/src/proof.rs

```diff
@@ -0,0 +1,102 @@
+use halo2_base::{
+    gates::{GateChip, GateInstructions},
+    poseidon::hasher::PoseidonHasher,
+    utils::BigPrimeField,
+    AssignedValue, Context,
+};
+
+use crate::poseidon_consts::{RATE, T};
+
+#[derive(Clone, Debug)]
+pub struct MerkleProof<F: BigPrimeField, const MAX_PATH_LEN: usize> {
+    pub path_shape: Vec<bool>,
+    pub path: Vec<F>,
+}
+
+#[derive(Clone, Debug)]
+pub struct CircuitMerkleProof<F: BigPrimeField, const MAX_PATH_LEN: usize> {
+    pub path_shape: Vec<AssignedValue<F>>,
+    pub path: Vec<AssignedValue<F>>,
+}
+
+impl<F: BigPrimeField, const MAX_PATH_LEN: usize> MerkleProof<F, MAX_PATH_LEN> {
+    pub fn new(path_shape: Vec<bool>, path: Vec<F>) -> Self {
+        Self { path_shape, path }
+    }
+
+    pub fn load(&self, ctx: &mut Context<F>) -> CircuitMerkleProof<F, MAX_PATH_LEN> {
+        let mut path_shape = vec![];
+        let mut path = vec![];
+
+        for i in 0..MAX_PATH_LEN {
```

Comment:
I'm not sure what happens here. Shouldn't `MAX_PATH_LEN` be equal to `self.path.len()`?

Reply:
not necessarily - we either have to pad data explicitly (which would allow us to convert the field from vec to array) or implicitly like this

Reply:
I think there is no point in padding. Also this is now incorrect, because padding is not taken into account in `verify`

Reply:
Done